### PR TITLE
Add token as parameter + gas saving with param; Add event EndVesting with total amount and Vest with amount.

### PR DESCRIPTION
**Pull request by alexisgayte**
- moved calculation on the top
- replace !=0 by >0
- cast timestamp
- delete before mint()
- add comment

- Add testToken/MkrTokenAuthority for testing.

- I like the real chain overridden But I am not too sure where to put it ...
Shouldn't it xxxxx.it.test

### CODE REVIEW

File: src/DssVest.sol [hunk 1]
```diff
@@ -52,46 +52,54 @@ contract DssVest {
...
-        if (_amt - _pmt != 0) {      // safe because pmt <= amt
+
+        if (_vestedAmount > 0) {
             awards[id] = Award({
                 usr: _usr,
                 bgn: uint48(block.timestamp),
-                fin: uint48(block.timestamp + _tau),
-                amt: uint128(_amt - _pmt),
+                fin: uint48(_endVesting),
+                amt: uint128(_vestedAmount),
```
Comment (brianmcmichael) [inline src/DssVest.sol:78]: We don't need to create a new variable for a value that's only used once.

File: src/DssVest.sol [hunk 2]
```diff
@@ -52,46 +52,54 @@ contract DssVest {
...
                 rxd: 0
             });
         }
-        if (_pmt != 0) {
+
+        emit Init(_usr, _amt, _endVesting);
```
Comment (brianmcmichael) [inline src/DssVest.sol:83]: Events should go at the end of a function. Not sure why you're doing it this way.

File: src/DssVest.sol [hunk 3]
```diff
@@ -52,46 +52,54 @@ contract DssVest {
...
 
     function init(address _usr, uint256 _amt, uint256 _tau, uint256 _pmt) external auth returns (uint256 id) {
         require(_usr != address(0),  "dss-vest/invalid-user");
-        require(_amt < uint128(-1),  "dss-vest/amount-error");
+        require(_amt < uint128(-1),  "dss-vest/amount-error"); // 1
         require(_tau < 5 * 365 days, "dss-vest/tau-too-long");
-        require(_pmt < uint128(-1),  "dss-vest/payout-error");
-        require(_pmt <= _amt,        "dss-vest/bulk-payment-higher-than-amt");
+        require(_pmt < uint128(-1),  "dss-vest/payout-error"); // 2
+        require(_pmt <= _amt,        "dss-vest/bulk-payment-higher-than-amt"); // 3 - 123 should be covered by math sub() for a better gas efficiency, Am I right?
```
Comment (brianmcmichael) [inline src/DssVest.sol:67]: This provides a clear error message to the user why the transaction is failing. The safemath reverts without a message.

File: src/DssVest.sol [hunk 4]
```diff
@@ -52,46 +52,54 @@ contract DssVest {
...
         wards[msg.sender] = 1;
         emit Rely(msg.sender);
+
+        MKR = IMKR(token);
     }
 
     function init(address _usr, uint256 _amt, uint256 _tau, uint256 _pmt) external auth returns (uint256 id) {
         require(_usr != address(0),  "dss-vest/invalid-user");
-        require(_amt < uint128(-1),  "dss-vest/amount-error");
+        require(_amt < uint128(-1),  "dss-vest/amount-error"); // 1
```
Comment (brianmcmichael) [inline src/DssVest.sol:64]: Not sure what these extra comments are here.

File: src/DssVest.sol [hunk 5]
```diff
@@ -52,46 +52,54 @@ contract DssVest {
     uint256 public ids;
 
     // Governance must rely() this contract on MKR to mint.
-    constructor() public {
+    constructor(address token) public {
```
Comment (brianmcmichael) [inline src/DssVest.sol:55]: This adds extra unnecessary complexity to a purpose-built contract that doesn't need it.

File: src/DssVest.sol [hunk 6]
```diff
@@ -52,46 +52,54 @@ contract DssVest {
...
     }
 
     function vest(uint256 _id) external {
         Award memory _award = awards[_id];
         require(_award.usr == msg.sender, "dss-vest/only-user-can-claim");
 
         if (block.timestamp >= _award.fin) {  // Vesting period has ended.
-            MKR.mint(_award.usr, _award.amt - _award.rxd); // TODO safemath
             delete awards[_id];
+            MKR.mint(_award.usr, _award.amt - _award.rxd); // TODO safemath
```
Comment (brianmcmichael) [inline src/DssVest.sol:96]: Minting should not happen after the mapping is deleted.

File: src/DssVest.sol [hunk 7]
```diff
@@ -52,46 +52,54 @@ contract DssVest {
...
+            // TODO I would add a new type of event and return.
         } else {                              // Vesting in progress
-            uint256 t = (block.timestamp - _award.bgn) * WAD / (_award.fin - _award.bgn);
-            uint256 mkr = (_award.amt * t) / WAD;
+            uint256 t = (uint48(block.timestamp) - _award.bgn) * WAD / (_award.fin - _award.bgn); // why do we mult by WAD? Do we need WAD here?
```
Comment (brianmcmichael) [inline src/DssVest.sol:99]: We multiply by WAD for additional precision here.

File: src/DssVest.sol [hunk 8]
```diff
@@ -52,46 +52,54 @@ contract DssVest {
...
+            uint256 mkr = (_award.amt * t) / WAD; // is * safe ? can it goes over? I am not too sure
+            awards[_id].rxd = uint128(mkr); // Do we need to mix uint128 with uint 256?
```
Comment (brianmcmichael) [inline src/DssVest.sol:101]: Using 128 allows us to pack the struct for gas savings.

File: src/DssVest.t.sol [hunk 9]
```diff
@@ -21,28 +23,28 @@ interface Token {
...
-        // Set testing contract as a MKR Auth
-        hevm.store(
-            address(GOV_GUARD),
-            keccak256(abi.encode(address(vest), uint256(1))),
-            bytes32(uint256(1))
-        );
-        assertEq(GovGuard(GOV_GUARD).wards(address(vest)), 1);
+
+        mkr = new TestToken("MKR", 18);
+        TokenAuthority mkrAuthority = new TokenAuthority();
```
Comment (brianmcmichael) [inline src/DssVest.t.sol:38]: The existing method of testing is less complex and tests against the actual token.

File: src/testhelper/MkrTokenAuthority.sol [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        if (sig == burn || sig == burnFrom || src == root) {
+            return true;
+        } else if (sig == mint) {
+            return (wards[src] == 1);
+        } else {
+            return false;
+        }
+    }
+}
```
Comment (brianmcmichael) [inline src/testhelper/MkrTokenAuthority.sol:40]: Always add newline at the end of files.

File: src/testhelper/TestToken.sol [hunk 11]
```diff
@@ -0,0 +1,7 @@
+pragma solidity 0.6.7;
+
+import "ds-token/token.sol";
+
+contract TestToken is DSToken {
+    constructor(bytes32 symbol_, uint256 decimals_) public DSToken(symbol_) {decimals = decimals_;}
+}
```
Comment (brianmcmichael) [inline src/testhelper/TestToken.sol:7]: newline.

**Review by brianmcmichael — COMMENTED** [review]
I'm not sure that this is an improvement. This adds unnecessary complexity and dependencies.

File: src/DssVest.sol (hunk 5 above)
Comment (alexisgayte) [reply src/DssVest.sol:55]: I thought it could be Dai too.

File: src/DssVest.sol (hunk 4 above)
Comment (alexisgayte) [reply src/DssVest.sol:64]: 1 /2 and 3 should be sort of managed by sub() + check the (result()  < uint128(-1) )
which I thought will be cleared and less call. But following you first comment I will remove it

File: src/DssVest.sol (hunk 1 above)
Comment (alexisgayte) [reply src/DssVest.sol:78]: _endVesting is used twice. I am actually not sure if it saves some gas.

File: src/DssVest.sol (hunk 6 above)
Comment (brianmcmichael) [reply src/DssVest.sol:96]: Check out how we do this in `deal`. Using the same pattern here.
https://github.com/makerdao/dss/blob/master/src/flop.sol#L160-L165

File: src/DssVest.sol (hunk 8 above)
Comment (alexisgayte) [reply src/DssVest.sol:101]: Sorry that was more a WIP stuff but thanks for the clarification.

File: src/DssVest.sol (hunk 2 above)
Comment (alexisgayte) [reply src/DssVest.sol:83]: https://github.com/BellwoodStudios/dss-psm/blob/master/MakerDAO%20Dai%20Peg%20Stability%20Module%20-%20Final%20Report.pdf 

It is inside the quantstamp report, actually I have no idea why, just applied it.

image

File: src/DssVest.t.sol (hunk 9 above)
Comment (alexisgayte) [reply src/DssVest.t.sol:38]: I actually like the way you do it. 
Tho it is more integration test, and more complex to set up.

File: src/DssVest.sol (hunk 6 above)
Comment (alexisgayte) [reply src/DssVest.sol:103]: it is actually different as you use the actual structure to pass the parameter to mint. 
        gem.mint(bids[id].guy, bids[id].lot);
        delete bids[id];

Comment (alexisgayte) [reply src/DssVest.sol:103]: gem can't be any token here. as mint() can call back deal() and mint as much as you call, I believe.

File: src/DssVest.t.sol (hunk 9 above)
Comment (alexisgayte) [reply src/DssVest.t.sol]: rolled back to it.

File: src/DssVest.sol (hunk 5 above)
Comment (alexisgayte) [reply src/DssVest.sol]: Took your code instead

**Comment by gbalabasquer** [comment]
@alexisgayte hope you don't mind but I'm going to close this PR in order to start cleaning the pending tasks.
The base contract was already changed regarding the version you branched off and will have further changes that the team has discussed. If you want to review it after we apply all the other PRs and still have suggestions please let us know.
Anyway this is much appreciated.

**Comment by alexisgayte** [comment]
Thanks appreciate the reply effort ;)

**Pull request closed without merging**